### ISSUE TITLE
TypeScript template should generate resolvers for directives, unions and interfaces

### ISSUE THREAD

**Issue by forabi** [dotansimha/graphql-code-generator#579]
The only reason I have a custom template is because I want to ensure type safety for these resolvers. If it could be included in the TS template, I could remove a lot of code and scripts in my codebase!

Here's how I generate these resolver types currently:

```typescript

/**
 * A union resolver tells GraphQL which one of the types of this union you are returning.
 *
 * > When you have a field in your schema that returns a union or interface type,
 * > you will need to specify an extra `__resolveType` field in your resolver map, which tells
 * > the GraphQL executor which type the result is, out of the available options.
 * @see https://www.apollographql.com/docs/graphql-tools/resolvers.html#Unions-and-interfaces
 */
type UnionOrInterfaceResolver<UnionOrInterface, PossibleTypeName, Context> = (
  object: UnionOrInterface,
  context: Context,
  info: GraphQLResolveInfo,
) => PossibleTypeName;

type BuiltInDirectives = 'pick' | 'include' | 'deprecated';

type CustomDirectiveMap = {
  [DirectiveName in Exclude<
    keyof DirectiveMap,
    BuiltInDirectives
  >]: DirectiveMap[DirectiveName]
};

export type DirectiveResolverMap = {
  [DirectiveName in keyof CustomDirectiveMap]: DirectiveResolver<
    DirectiveMap[DirectiveName]['args'],
    SchemaContext
  >
};

export type UnionResolverMap = {
  [UnionName in keyof UnionMap]: {
    __resolveType: UnionOrInterfaceResolver<
      UnionMap[UnionName]['unionType'],
      UnionMap[UnionName]['possibleTypeNames'],
      SchemaContext
    >;
  }
};

export type InterfaceResolverMap = {
  [InterfaceName in keyof InterfaceMap]: {
    __resolveType: UnionOrInterfaceResolver<
      InterfaceMap[InterfaceName]['interfaceType'],
      InterfaceMap[InterfaceName]['implementorTypeNames'],
      SchemaContext
    >;
  }
};

```

```handlebars

// region Directives

export type DirectiveMap = {
  {{#each definedDirectives}}
  /** {{ description }} */
  {{ name }}: {
    args: {{# if hasArguments }}{{ toPascalCase name }}{{ toPascalCase ../name }}Args{{else}}{ }{{/if}};
  }
  {{/each}}
};

export type UnionMap = {
  {{#each unions}}
  {{ name }}: {
    unionType: {{name}};
    possibleTypeNames: {{#each possibleTypes}}'{{this}}'{{#unless @last}} | {{/unless}}{{/each}};
  }
  {{/each}}
}

export type InterfaceMap = {
  {{#each interfaces}}
  {{ name }}: {
    interfaceType: {{name}};
    implementorTypeNames: {{#if hasImplementingTypes}}{{#each implementingTypes}}'{{this}}'{{#unless @last}} | {{/unless}}{{/each}}{{else}}never{{/if}};
  }
  {{/each}}
}

{{#each definedDirectives}}
    {{#if hasArguments }}
export interface {{ toPascalCase name }}Args {
{{#each arguments}}

  /** {{ description }} */
  {{ name }}{{#unless isRequired}}?{{/unless}}: {{ toPrimitive type }}{{#if isArray}}[]{{/if}}{{#unless isRequired}} | null{{/unless}};
{{/each}}
}
    {{/if}}
{{/each}}

// endregion Directives

```

**Comment by dotansimha** [comment]
Sounds good. @kamilkisiela what do you think?
@forabi do you want to create a PR for it?

**Comment by dotansimha** [comment]
@forabi implemented in: https://github.com/dotansimha/graphql-code-generator/pull/910

**Comment by dotansimha** [comment]
Fixed in 0.14.2 🎉
